jwt/jwt.go

```diff
@@ -0,0 +1,265 @@
+		latestStart := *claims.IssuedAt
+		if *claims.NotBefore > *claims.IssuedAt {
+			latestStart = *claims.NotBefore
+		}
+		leeway := expected.ExpirationLeeway.Seconds()
+		if expected.ExpirationLeeway.Seconds() < 0 {
+			leeway = 0
+		} else if expected.ExpirationLeeway.Seconds() == 0 {
+			leeway = DefaultLeewaySeconds
+		}
+		*claims.Expiry = jwt.NumericDate(int64(latestStart) + int64(leeway))
+	}
+
+	// If "nbf" (Not Before) is not set, then set it to the "iat" (Issued At) if set.
+	// Otherwise, set it to the "exp" (Expiration Time) minus leeway.
+	if *claims.NotBefore == 0 {
+		if *claims.IssuedAt != 0 {
+			*claims.NotBefore = *claims.IssuedAt
+		} else {
+			leeway := expected.NotBeforeLeeway.Seconds()
+			if expected.NotBeforeLeeway.Seconds() < 0 {
+				leeway = 0
+			} else if expected.NotBeforeLeeway.Seconds() == 0 {
+				leeway = DefaultLeewaySeconds
+			}
+			*claims.NotBefore = jwt.NumericDate(int64(*claims.Expiry) - int64(leeway))
+		}
+	}
+
+	// Set clock skew leeway to apply when validating all time-related claims
+	cksLeeway := expected.ClockSkewLeeway
+	if expected.ClockSkewLeeway.Seconds() < 0 {
+		cksLeeway = 0
+	} else if expected.ClockSkewLeeway.Seconds() == 0 {
+		cksLeeway = jwt.DefaultLeeway
+	}
+
+	// Validate claims by asserting they're as expected
+	if expected.Issuer != "" && expected.Issuer != claims.Issuer {
+		return nil, fmt.Errorf("invalid issuer (iss) claim")
+	}
+	if expected.Subject != "" && expected.Subject != claims.Subject {
+		return nil, fmt.Errorf("invalid subject (sub) claim")
+	}
+	if expected.ID != "" && expected.ID != claims.ID {
+		return nil, fmt.Errorf("invalid ID (jti) claim")
+	}
+	if err := validateAudience(expected.Audiences, claims.Audience); err != nil {
+		return nil, fmt.Errorf("invalid audience (aud) claim: %w", err)
+	}
+
+	// Validate that the token is not expired with respect to the current time
+	now := time.Now()
+	if expected.Now != nil {
+		now = expected.Now()
+	}
+	if claims.NotBefore != nil && now.Add(cksLeeway).Before(claims.NotBefore.Time()) {
+		return nil, errors.New("invalid not before (nbf) claim: token not yet valid")
+	}
+	if claims.Expiry != nil && now.Add(-cksLeeway).After(claims.Expiry.Time()) {
+		return nil, errors.New("invalid expiration time (exp) claim: token is expired")
+	}
+	if claims.IssuedAt != nil && now.Add(cksLeeway).Before(claims.IssuedAt.Time()) {
+		return nil, errors.New("invalid issued at (iat) claim: token issued in the future")
+	}
+
+	return allClaims, nil
+}
+
+// validateSigningAlgorithm checks whether the JWS "alg" (Algorithm) header
+// parameter value for the given JWT matches any given in expectedAlgorithms.
+// If expectedAlgorithms is empty, RS256 will be expected by default.
+func validateSigningAlgorithm(token string, expectedAlgorithms []Alg) error {
+	for _, expected := range expectedAlgorithms {
+		if !supportedAlgorithms[expected] {
+			return fmt.Errorf("unsupported signing algorithm %q", expected)
+		}
+	}
```

Comment:
Is this check something we'll be able to run at config time too?

Reply:
I've introduced a function in 72a2a3b that'll allow us to run this check at config time.